drivers/network/dd/b57xx/interrupt.c

```diff
@@ -0,0 +1,185 @@
+            ProducerIndex = Adapter->StandardProducer.Index;
+            pReceiveDescriptor = Adapter->ReturnConsumer[0].pRing + ConsumerIndex;
+            
+            if (pReceiveDescriptor->Flags & B57XX_RBD_FRAME_HAS_ERROR)
+            {
+                NDIS_MinDbgPrint("RX error (0x%x)\n", pReceiveDescriptor->Flags);
+                Adapter->Statistics.ReceiveErrors++;
+                goto ContinueReceive;
+            }
+            
+            BufferBase = (PCHAR)Adapter->StandardProducer.HostBuffer +
+                         Adapter->StandardProducer.FrameBufferLength * pReceiveDescriptor->Index;
+            
+            NdisDprReleaseSpinLock(&Adapter->Interrupt.Lock);
+            NdisMEthIndicateReceive(Adapter->MiniportAdapterHandle,
+                                    NULL,
+                                    BufferBase,
+                                    sizeof(ETH_HEADER),
+                                    BufferBase + sizeof(ETH_HEADER),
+                                    pReceiveDescriptor->Length - sizeof(ETH_HEADER),
+                                    pReceiveDescriptor->Length - sizeof(ETH_HEADER));
+            NdisDprAcquireSpinLock(&Adapter->Interrupt.Lock);
+            GotAny = TRUE;
+            
+            Adapter->Statistics.ReceiveSuccesses++;
+
+ContinueReceive:
+            ConsumerIndex = (ConsumerIndex + 1) % Adapter->ReturnConsumer[0].Count;
+            Adapter->ReturnConsumer[0].Index = ConsumerIndex;
+            
+            ProducerIndex = (ProducerIndex + 1) % Adapter->StandardProducer.Count;
+            Adapter->StandardProducer.Index = ProducerIndex;
+        }
+        
+        if (GotAny)
+        {
+            NICReceiveSignalComplete(Adapter);
+            NdisDprReleaseSpinLock(&Adapter->Interrupt.Lock);
+            NdisMEthIndicateReceiveComplete(Adapter->MiniportAdapterHandle);
+            NdisDprAcquireSpinLock(&Adapter->Interrupt.Lock);
+        }
+        
+        /* Process any completed frames (transmit interrupts) */
+        while (Adapter->SendProducer[0].ConsumerIndex !=
+               Adapter->Status.pBlock->RingIndexPairs[0].SendConsumerIndex)
+        {
+            ConsumerIndex = Adapter->SendProducer[0].ConsumerIndex;
+            
+            NdisDprReleaseSpinLock(&Adapter->Interrupt.Lock);
+            NdisMSendComplete(Adapter->MiniportAdapterHandle,
+                             Adapter->SendProducer[0].pPacketList[ConsumerIndex],
+                             NDIS_STATUS_SUCCESS);
+            NdisDprAcquireSpinLock(&Adapter->Interrupt.Lock);
+            Adapter->SendProducer[0].RingFull = FALSE;
+            
+            Adapter->Statistics.TransmitSuccesses++;
+            
+            ConsumerIndex = (ConsumerIndex + 1) % Adapter->SendProducer[0].Count;
+            Adapter->SendProducer[0].ConsumerIndex = ConsumerIndex;
+        }
+    }
+    while (NICInterruptCheckAvailability(Adapter));
```

Comment:
You should avoid doing too much work at DPC level. You could count the number of inner loop and then exit the DPC if it reaches some limit.